### PR TITLE
[Alert manager] k8s cert expiration checker

### PR DESCRIPTION
**Pull request by yiyione**
#5334

### CODE REVIEW

File: src/alert-manager/deploy/service.yaml [hunk 1]
```diff
@@ -37,4 +38,4 @@ upgraded-script: upgraded.sh
 
 
 deploy-rules:
-  - in: pai-master
\ No newline at end of file
+  - in: pai-master
```
Comment (suiguoxin) [inline src/alert-manager/deploy/service.yaml:41]: The `deploy-rule` doesn't apply for CronJob, https://github.com/microsoft/pai/blob/master/deployment/paiLibrary/paiService/service_template_generate.py#L61
 we can add a nodeSelector in the CronJob definition.

File: src/alert-manager/config/alert-manager.yaml [hunk 2]
```diff
@@ -31,3 +31,6 @@ cluster-utilization:
   configured: False
 use-pylon: False
 repeat-interval: '24h'
+cert-expiration-checker:
+  schedule: '0 0 * * *'
+  alert-residual-days: 30
```
Comment (suiguoxin) [inline src/alert-manager/config/alert-manager.yaml:36]: This file is loaded as the default config, this part should be configurable in `services-configuration.yaml`, refer to #5281

File: src/alert-manager/src/cert-expiration-checker/send_alert.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+def send_alert(pai_url: str, certExpirationInfo: str):
+    trigger_time = str(datetime.now(timezone.utc).date())
+    post_url = pai_url.rstrip("/") + ALERT_PREFIX
+    alerts = []
+    alert = {
+        "labels": {
+            "alertname": "k8s cert expiration",
+            "severity": "warn",
+            "trigger_time": trigger_time,
+            "annotations": certExpirationInfo,
```
Comment (suiguoxin) [inline src/alert-manager/src/cert-expiration-checker/send_alert.py:34]: Please provide a detailed summary in the `annotations` field like:
```yaml
annoations:
    summary: the k8s cert will be expired in {} days
``` 
The summary will be rendered in the email body

File: src/alert-manager/deploy/alert-manager-cert-expiration-check-cronjob.yaml.template [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    {%- else %}
+              value: "{{ cluster_cfg['pylon']['uri']}}"
+    {%- endif %}
+            - name: ALERT_RESIDUAL_DAYS
+              value: "{{ cluster_cfg["alert-manager"]["cert-expiration-checker"]["alert-residual-days"] }}"
+            volumeMounts:
+            - mountPath: /usr/local/bin/kubeadm
+              name: kubeadm
+            - mountPath: /etc/kubernetes/
+              name: kubenetes-config
```
Comment (Binyang2014) [inline src/alert-manager/deploy/alert-manager-cert-expiration-check-cronjob.yaml.template:31]: Can make the path configurable? We can't assume all env such as AKS use this path.

File: src/alert-manager/src/cert-expiration-checker/send_alert.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    alerts.append(alert)
+
+    logging.info("Sending alerts to alert-manager...")
+    resp = requests.post(post_url, json=alerts)
+    resp.raise_for_status()
+    logging.info("Alerts sent to alert-manager.")
+
+def main():
+    PAI_URI = os.environ.get("PAI_URI")
+    certExpirationInfo = os.popen('kubeadm alpha certs check-expiration --config="/etc/kubernetes/kubeadm-config.yaml"').read()
```
Comment (Binyang2014) [inline src/alert-manager/src/cert-expiration-checker/send_alert.py:47]: The command seems depends on kubeadm version. Refer https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#check-certificate-expiration, the latest version don't need to add alpha in command.

One way to solve this is package kubeadm into docker image
Another way is check the cert manually. It's easy to use a lib to check cert expired time

File: src/alert-manager/deploy/alert-manager-cert-expiration-check-cronjob.yaml.template (hunk 4 above)
Comment (yiyione) [reply src/alert-manager/deploy/alert-manager-cert-expiration-check-cronjob.yaml.template:31]: OK, I'll change it

File: src/alert-manager/config/alert-manager.yaml (hunk 2 above)
Comment (yiyione) [reply src/alert-manager/config/alert-manager.yaml:36]: Done.

File: src/alert-manager/src/cert-expiration-checker/send_alert.py (hunk 3 above)
Comment (yiyione) [reply src/alert-manager/src/cert-expiration-checker/send_alert.py]: Done.

File: src/alert-manager/src/cert-expiration-checker/send_alert.py (hunk 5 above)
Comment (yiyione) [reply src/alert-manager/src/cert-expiration-checker/send_alert.py]: Done. use python OpenSSL.crypto.load_certificate(...) to read the crt file, and check the expired time.

**Comment by suiguoxin** [comment]
Could you provide some documents to the end of this file : https://github.com/microsoft/pai/blob/master/docs/manual/cluster-admin/how-to-use-alert-system.md

**Comment by yiyione** [comment]
> Could you provide some documents to the end of this file : https://github.com/microsoft/pai/blob/master/docs/manual/cluster-admin/how-to-use-alert-system.md

Added.

**Review by suiguoxin — APPROVED** [review]

**Comment by siaimes** [comment]
> > Could you provide some documents to the end of this file : https://github.com/microsoft/pai/blob/master/docs/manual/cluster-admin/how-to-use-alert-system.md
> 
> Added.

> This feature will be enable by default, if the action `email-admin` is enabled.
> You can configure the `alert-manager`->`cert-expiration-checker` field in `services-configuration.yml`.
> `schedule`, `alert-residual-days` and `cert-path` are necessary fields for this feature, and we have default value for the fields.

Hi, @yiyione 

I have some doubts about this description. Is the warning message not sent to the upper right corner of the homepage? If the warning message is sent to the upper right corner of the homepage, can I use this feature without setting up a mail server?

Thanks.

**Comment by siaimes** [comment]
[kubernetes/kubernetes/pull/52196](https://github.com/kubernetes/kubernetes/pull/52196)

In addition, I found a PR in the kubernetes repository that seems to fix the problem. Is this PR in this repository still necessary?

**Comment by Binyang2014** [comment]
@siaimes Nice catch! We don't notice this feature. For this release, we will still use this PR. After we verified this feature, this PR can revert. Thanks for this notice!

**Comment by yiyione** [comment]
> Hi, @yiyione
> 
> I have some doubts about this description. Is the warning message not sent to the upper right corner of the homepage? If the warning message is sent to the upper right corner of the homepage, can I use this feature without setting up a mail server?
> 
> Thanks.

It's a good question, and we are going to fix it.
After the PR #5440 merged, the admin will be able to view the alert on the `upper right corner of the homepage` without setting up the email.

Thanks.

**Pull request merged**